Enhancement for smart serialization for operator schemas with out arg

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63651
* #63540
* __->__ #63096



Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

tugsbayasgalan committed:
Enhancement for smart serialization for out schemas

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




[ghstack-poisoned]

raziel commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1180,6 +1182,28 @@ struct PythonPrintImpl {
         }
+
+        // scan backwards to find the first index where out args start
+        size_t out_start = num_schema_args - 1;
+        while (out_start >= necessary_args) {
+          auto arg = schema.arguments().at(out_start);
+          if (!arg.is_out()) {
+            break;
+          }
+          out_start--;
+        }
+
+        // print out args
+        for (size_t i = out_start + 1; i < schema.arguments().size(); i++) {
```
Small suggestion:

```suggestion
        size_t out_start = num_schema_args;
        while (--out_start >= necessary_args) {
          auto arg = schema.arguments().at(out_start);
          if (!arg.is_out()) {
            break;
          }
        }

        // print out args
        for (size_t i = out_start; i < schema.arguments().size(); i++) {
```

That way ```out_start``` really ends up with the right index, and the for loop can start from that point without adjusting.

tugsbayasgalan replied:
thanks :) will do

raziel commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1180,6 +1182,28 @@ struct PythonPrintImpl {
+          TORCH_INTERNAL_ASSERT(arg.is_out());
+          // figure out the corresponding input at this index
+          auto input_idx = node->inputs().size() - (num_schema_args - i);
+          if (input_idx < node->inputs().size()) {
```
why is this IF check needed?

tugsbayasgalan replied:
I think there could be cases where actual provided inputs are less/more than the number of arguments in the schema due to star sign in the schema. In that case, I think there will index mismatch so only reliable way to tell out args from the input would be assuming end inputs are out.

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




[ghstack-poisoned]

cccclai commented on torch/csrc/jit/runtime/calculate_necessary_args.h:
```diff
@@ -13,9 +13,19 @@ inline size_t CalculateNecessaryArgs(
     return actual_inputs.size();
   }
+
+  int schema_idx = schema_args.size() - 1;
```
@iseeyuan what should be the expected `op_to_num_specified_args` with this change? For the following source code
```
class Module(torch.nn.Module):
    def __init__(self):
        super().__init__()
    def forward(self, x, h):
        torch.add(x, h, alpha=2.0, out=x)
```

Should bytecode looks like 
a. 
```
...
('operators', (('aten::add', 'out', 3),))
...
  (('arguments',
    ((('name', 'self'), ('type', '__torch__.m'), ('default_value', None)),
     (('name', 'x'), ('type', 'Tensor'), ('default_value', None)),
     (('name', 'h'), ('type', 'Tensor'), ('default_value', None)))),
```
or 
```
...
('operators', (('aten::add', 'out', 4),))
...
  (('arguments',
    ((('name', 'self'), ('type', '__torch__.m'), ('default_value', None)),
     (('name', 'x'), ('type', 'Tensor'), ('default_value', None)),
     (('name', 'h'), ('type', 'Tensor'), ('default_value', None)))),
```

With this pr, looks like the `op_to_num_specified_args` is 3 now.

iseeyuan replied:
I think it should be 3. In the .cpp test, if `alpha = 1.0`, `op_to_num_specified_args` = 2.  @tugsbayasgalan could you confirm?

cccclai replied:
I'm checking the bytecode from the source code above, with and without this pr:

With this pr:
![image](https://user-images.githubusercontent.com/16430979/129623334-a6375f00-8a4e-40de-bcb0-8580c497990a.png)


Without this pr:
![image](https://user-images.githubusercontent.com/16430979/129623348-2bb15d3c-0a97-43e5-bdce-d1e66f9e00e3.png)


There are two main differences:
1. the number of specified arguments is different 4->3
2. The instructions are different, we no longer have the `('Move', 2, 0)`. Is it expected?

Another question, for `aten::add` operator, there are four arguments in total. Is it expected? What should be the correct sequence?

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




[ghstack-poisoned]

tugsbayasgalan commented on torch/csrc/jit/runtime/interpreter/code_impl.h:
```diff
@@ -737,8 +737,8 @@ struct MobileCodeImpl : CodeImpl {
         if (!op_schema.is_vararg()) {
-          auto numInclude =
-              CalculateNecessaryArgs(op_schema.arguments(), node->inputs());
+          auto numInclude = CalculateNecessaryArgs(
```
@cccclai You will need to flip this boolean after you are done with your change.

cccclai replied:
I check two cases:
```
torch.add(x, h, alpha=1.0, out=x)
```
actual output: `('aten::add', 'out', 4)`, 
expected output: `('aten::add', 'out', 3)`, because alpha default is 3

This one is correct:
```
`torch.add(x, h, out=x) => ('aten::add', 'out', 3)`
```
I will use `torch.add(x, h, out=x)` for now to test, can you help checking the first case?

tugsbayasgalan replied:
Right now its 4 because the flag is set to false, so its following the original code path. Could u elaborate what you mean by the first case?

cccclai replied:
I set the flag to true, but `torch.add(x, h, alpha=1.0, out=x)` produces the result `('aten::add', 'out', 4)`. In the meanwhile, `torch.add(x, h, out=x)` produces the correct result `('aten::add', 'out', 3)`. Does it make sense?

tugsbayasgalan replied:
Hmm are you sure? It still produces 3 on my local machine. The example graph I used was:
```
R"IR(
    graph(%x.1 : Tensor,
          %y.1 : Tensor):
      %8 : NoneType = prim::Constant()
      %6 : int = prim::Constant[value=1]()
      %7 : Tensor = aten::add(%x.1, %y.1, %6, %y.1)
      return (%8))IR";
```

cccclai replied:
Thanks for the follow up. The default value should be 1 instead of 1.0.  `torch.add(x, h, alpha=1, out=x)` can produce the correct output: 3.

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




[ghstack-poisoned]

tugsbayasgalan commented on torch/csrc/jit/runtime/interpreter/code_impl.h:
```diff
@@ -737,13 +745,19 @@ struct MobileCodeImpl : CodeImpl {
+          auto specifiedArgs = CalculateNecessaryArgs(
+              op_schema.arguments(), node->inputs(), false);
+          // preserving the old behavior
+          auto numInclude = specifiedArgs.first;
```
@cccclai  can you comment this out and uncomment the next lines?

tugsbayasgalan commented on torch/csrc/jit/runtime/interpreter/code_impl.h:
```diff
@@ -769,6 +783,13 @@ struct MobileCodeImpl : CodeImpl {
           num_include = it->second;
         }
         emitLoadInputs(node->inputs(), num_include);
```
comment this one out and uncomment the next chunk

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415255).

Commit:
Update on "Enhancement for smart serialization for out schemas"


Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

Commit:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

Commit:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

Commit:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

Commit:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

Commit:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

Commit:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415255).

Commit:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1162,23 +1162,67 @@ struct PythonPrintImpl {
+            schema_idx--;
+          }
+
+          auto num_out = schema.arguments().size() - schema_idx - 1;
```
Is this different from `specified_args.second`?

tugsbayasgalan replied:
hhha good catch. It is the same thing

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1162,23 +1162,67 @@ struct PythonPrintImpl {
+          }
+
+          auto num_out = schema.arguments().size() - schema_idx - 1;
+          auto num_specified_before_out = necessary_args - num_out;
```
Is this `specified_args.first`?

tugsbayasgalan replied:
good catch. It is the same thing

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415255).

tugsbayasgalan commented on torch/csrc/jit/runtime/calculate_necessary_args.h:
```diff
@@ -7,18 +7,42 @@
 namespace jit {
 
-inline size_t CalculateNecessaryArgs(
+inline std::pair<size_t, size_t> CalculateNecessaryArgs(
```
thanks will do

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415255).

Commit:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415255).

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415255).

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415255).

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415255).

tugsbayasgalan commented on torch/csrc/jit/runtime/calculate_necessary_args.h:
```diff
@@ -9,16 +9,33 @@ namespace jit {
     const std::vector<Argument>& schema_args,
-    at::ArrayRef<Value*> actual_inputs) {
+    at::ArrayRef<Value*> actual_inputs,
+    bool add_out_args) {
```
done

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415255).

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415255).

Commit:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415255).

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

Commit:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

Commit:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

Commit:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]

Commit:
Update on "Enhancement for smart serialization for out schemas"




Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255/)

Differential Revision: [D30415255](https://our.internmc.facebook.com/intern/diff/D30415255)

[ghstack-poisoned]
